pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/RecordExtractor.java

```diff
@@ -0,0 +1,40 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.data.readers;
+
+import com.google.common.base.Preconditions;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.data.TimeFieldSpec;
+
+
+public abstract class RecordExtractor<T> {
+
+  protected FieldSpec getFieldSpecToUse(Schema schema, FieldSpec incomingFieldSpec) {
+    if (incomingFieldSpec.getFieldType() == FieldSpec.FieldType.TIME) {
+      TimeFieldSpec timeFieldSpec = schema.getTimeFieldSpec();
+      Preconditions.checkNotNull(timeFieldSpec);
+      return new TimeFieldSpec(timeFieldSpec.getIncomingGranularitySpec());
+    }
+
+    return incomingFieldSpec;
+  }
+
+  abstract public GenericRow extract(Schema schema, T from, GenericRow to);
```

Comment:
this interface assumes T is self-describing (like thrift, avro, json) but it fails for format like csv, tsv etc where we get a comma separated line. This means it needs an init method :(.

Reply:
Can we add one when necessary? Should not be too hard to add.

Reply:
Also, it is not necessary for decoders to use record extractors. If it is trivial they can choose to extract the record with a statement like "x = record.field"